internal/pkg/policy/policy_output.go

```diff
@@ -97,6 +111,64 @@ func (p *Output) prepareElasticsearch(
 		agent.Outputs[p.Name] = output
 	}
 
+	// retire api key of removed remote output
+	var toRetireAPIKeys *model.ToRetireAPIKeyIdsItems
+	var removedOutputKey string
+	// find the first output that is removed - supposing one output can be removed at a time
+	for agentOutputKey, agentOutput := range agent.Outputs {
+		found := false
+		for outputMapKey := range outputMap {
+			if agentOutputKey == outputMapKey {
+				found = true
+			}
+		}
+		if !found {
+			zlog.Info().Str("APIKeyID", agentOutput.APIKeyID).Str("output", agentOutputKey).Msg("Output removed, will retire API key")
+			toRetireAPIKeys = &model.ToRetireAPIKeyIdsItems{
+				ID:        agentOutput.APIKeyID,
+				RetiredAt: time.Now().UTC().Format(time.RFC3339),
+				Output:    agentOutputKey,
+			}
+			removedOutputKey = agentOutputKey
+			break
+		}
+	}
+
+	if toRetireAPIKeys != nil {
+
+		// adding remote API key to new output toRetireAPIKeys
+		fields := map[string]interface{}{
+			dl.FieldPolicyOutputToRetireAPIKeyIDs: *toRetireAPIKeys,
+		}
+
+		// Using painless script to append the old keys to the history
+		body, err := renderUpdatePainlessScript(p.Name, fields)
+		if err != nil {
+			return fmt.Errorf("could not update painless script: %w", err)
+		}
+
+		if err = bulker.Update(ctx, dl.FleetAgents, agent.Id, body, bulk.WithRefresh(), bulk.WithRetryOnConflict(3)); err != nil {
+			zlog.Error().Err(err).Msg("fail update agent record")
+			return fmt.Errorf("fail update agent record: %w", err)
+		}
+
+		// remove output from agent doc
+		body, err = json.Marshal(map[string]interface{}{
+			"script": map[string]interface{}{
+				"lang":   "painless",
+				"source": fmt.Sprintf("ctx._source['outputs'].remove(\"%s\")", removedOutputKey),
+			},
+		})
+		if err != nil {
+			return fmt.Errorf("could not create request body to update agent: %w", err)
+		}
+
+		if err = bulker.Update(ctx, dl.FleetAgents, agent.Id, body, bulk.WithRefresh(), bulk.WithRetryOnConflict(3)); err != nil {
+			zlog.Error().Err(err).Msg("fail update agent record")
```

Comment:
Could these `err` strings that we log contain the API key itself?

Reply:
good question, let me check

Reply:
tested (with an invalid script) and the API key is not logged:
```
10:47:23.146 ERR fail update agent record error.message="elastic fail 400: illegal_argument_exception: failed to execute script: script_exception: compile error" ecs.version=1.6.0 fleet.access.apikey.id=Pe2914sBdoeXGYc8n11i fleet.agent.id=a047dc88-9409-400a-b66b-3d83df8de26b fleet.ctx=processPolicy fleet.policy.id=9602dfd2-fc1d-423d-8607-e2b46b819560 fleet.policy.output.name=default fleet.policyCoordinator=1 fleet.policyRevision=40 http.request.id=01HFBVVVAHASQPJQ3XA92FG728 server.address= service.name=fleet-server service.type=fleet-server
```